### ISSUE TITLE
Allow loading the root app and child apps asynchronously and registering apps 

### ISSUE THREAD

**Issue by reaktivo** [Travix-International/frint#193]
## Currently

The root application is required to load and inject itself into the window global before any other child application can register. This prevents us from adding the async attribute to the script, which allows us to load multiple scripts in parallel.

```
// In index.html
<script src="https://example.com/assets/root.js"></script>
<script src="https://example.com/assets/childApp.js"></script>
```

``` 
// In root.js
window.app = new App();
render(window.app, containerElement);
```

```
// In childApp.js
window.app.registerApp(ChildApp, {
  regions: ['main'],
});
```

## Proposed

The proposed changes would allow to load root and child apps asynchronously, but in a way that execution order is not guaranteed. Doing this would increase the time to first paint and would prevent the script from blocking any other script parsing and/or layout render.

We achieve this by injecting a placeholder array for the root app on each child app, when it's not currently available. When the root app is loaded, we replace the placeholder array with the root app. By adding a `push` method to the Application API, we can emulate the array's own. This way, child apps don't need to be aware if they're pushing to an array or to the actual root application.

Prior Art:  http://googlecode.blogspot.nl/2009/12/google-analytics-launches-asynchronous.html 

```
// In index.html
// Note the addition of the async attribute
<script src="https://example.com/assets/rootApp.js" async></script>
<script src="https://example.com/assets/childApp.js" async></script>
```

``` 
// In root.js
const rootApp = new RootApp();
window.app.forEach(app => rootApp.push(app));
window.app = rootApp;
render(rootApp, containerElement);
```

```
// In childApp.js
window.app || window.app = [];
window.app.push({
  app: ChildApp,
  regions: ['main'],
});
```

## To be done:

- [ ] Add the `push` method to the App component, this should internally call the registerApp method.
- [ ] ~Remove the dependency on any window global in library code, see [Region.js](https://github.com/Travix-International/frint/blob/master/packages/frint-react/src/components/Region.js#L45)~ See Fahad's [comment](https://github.com/Travix-International/frint/issues/193#issuecomment-301056770)
- [ ] Create test scenarios where child apps run before the root app.

**Comment by fahad19** [comment]
cool stuff! I really like the proposal of making all `<script>`s async.

Replying to the `to be done` items below:

---

> Add the push method to the App component, this should internally call the registerApp method.

I am hesitant to add any new method at `App` class level in `frint` package. Since the proposal is a different approach than the docs, and a custom implementation of its own, may be we can do something like:

```js
// root-app/index.js
window.app = new RootApp();
window.app.push = function () { ... };
```

---

> Remove the dependency on any window global in library code, see Region.js

Since 1.x, we don't depend on globals any more. I believe this specific line was left out for `frint-compat`. But needs further investigating of course.

---

> Create test scenarios where child apps run before the root app.

👍 👍

**Comment by reaktivo** [comment]
@fahad19 Sounds reasonable, so no API changes. Do you agree things do be done then would be:

- Updating the documentation to include this use case
- Create or update examples to show this is possible.
- Create a test for this.

**Comment by fahad19** [comment]
I would love it if there is a new page here under `HOW TOS` for detailing this approach.

You can also create a new example for this in the repo. Something you may want to point to from the docs.

**Comment by fahad19** [comment]
Closed via #206.
